Enable Partition Discovery for Broker Load

yuanlihan opened the pull request:
Currently, we support loading data from parquet file, but can not parse partition columns in the path of parquet file and can not recursively list all files under base path of input.

This patch is able to discover and infer partitioning information under the base path of input like in Spark. It recursively list all the files under the base path and parse partition columns base on the base path if needed.

This patch parse partition columns at BrokerScanNode.java and save parsing result of each file path as a property of TBrokerRangeDesc, then  parquet_reader of BE can read the value of specified partition column.

yuanlihan committed:
Enable partition discovery when loading data from parquet file

yuanlihan committed:
Enable partition discovery when loading data from parquet file

morningman committed:
Fix bug that replicas of a tablet may be located on same host (#1517)

Doris support deploy multi BE on one host. So when allocating BE for replicas of
a tablet, we should select different host. But there is a bug in tablet scheduler
that same host may be selected for one tablet. This patch will fix this problem.

There are some places related to this problem:

1. Create Table
    There is no bug in Create Table process.

2. Tablet Scheduler
    Fixed when selecting BE for REPLICA_MISSING and REPLICA_RELOCATING.
    Fixed when balance the tablet.

3. Colocate Table Balancer
    Fixed when selecting BE for repairing colocate backend sequence.
    Not fix in colocate group balance. Leave it to colocate repairing.

4. Tablet report
    Tablet report may add replica to catalog. But I did not check the host here,
    Tablet Scheduler will fix it.

imay commented:
@yuanlihan 

Firstly, thanks for your improvement. I have some questions about this PR.

1. Should we add an option to switch this feature off? Because this will fill partition row automatically which should be known by users.

2. Can this patch work with other format of file? like CSV format? And how can user use this function.

3. How this function work with other load functions, like user specified columns and set operation. For example, if the partition column name is different in Doris table and Spark.

yuanlihan commented:
> @yuanlihan
> 
> Firstly, thanks for your improvement. I have some questions about this PR.
> 
> 1. Should we add an option to switch this feature off? Because this will fill partition row automatically which should be known by users.
> 2. Can this patch work with other format of file? like CSV format? And how can user use this function.
> 3. How this function work with other load functions, like user specified columns and set operation. For example, if the partition column name is different in Doris table and Spark.

@imay 
Thanks for your review
1. In my opinion, it is quite intuitive to enable partition discovery as well as recursively list files by default. We extract partition columns iff them was defined/needed in the specified table.
2. We can support partition discovery of other file sources(including Text/CSV/JSON/ORC/Parquet) like in Spark. May be users will prefer having this feature by default, which rarely have conflicts with their previous usage.
3. I will dig more about this and try to solve it.

imay committed:
Switch MAKE_TEST off in build.sh (#1579)

EmmyMiao87 committed:
Thrown TabletQuorumFailedException in commitTxn (#1575)

The TabletQuorumFailedException will be thrown in commitTxn while the success replica num of tablet is less then quorom replica num.
The Hadoop load does not handle this exception because the push task will retry it later.
The streaming broker, insert, stream and mini load will catch this exception and abort the txn after that.

yuanlihan committed:
Read date type column from parquet

yuanlihan committed:
Read date type column from parquet

yuanlihan committed:
Enable partition discovery for broker loading

yuanlihan committed:
Enable partition discovery for broker loading

yuanlihan commented:
Add Partition Discover as a feature of Broker Load.

For usage of this feature see diffs of related docs.

yuanlihan committed:
Update comment

yuanlihan committed:
Update comment

imay commented on docs/help/Contents/Data Manipulation/broker_load.md:
```diff
@@ -352,15 +357,47 @@
+        1. 不指定Partition Discovery的基础路径（BASE_PATH）
+            LOAD LABEL example_db.label10
+            (
+            DATA INFILE("hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir")
```
这样直接指定目录就可以导入，跟我们之前的默认行为是有冲突的。有的时候用户写错了，可能会将整个目录导入。
为了避免这种情况，可以不使用DATA INFILE，可是使用类似DATA INDIR的方式来指定导入一个目录

imay commented on docs/help/Contents/Data Manipulation/broker_load.md:
```diff
@@ -352,15 +357,47 @@
+            INTO TABLE `my_table`
+            FORMAT AS "csv"
+            BASE_PATH AS "hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir/"
+            (k1, k2, k3, utc_date，city)
```
这里columns表示的一般是文件中包含的列名，如果加上utc_date, city可能会让用户感到困惑。
建议不要再这个地方声明partition列名

imay commented on docs/help/Contents/Data Manipulation/broker_load.md:
```diff
@@ -352,15 +357,47 @@
+            DATA INFILE("hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir/city=beijing/utc_date=2019-06-26")
+            INTO TABLE `my_table`
+            FORMAT AS "csv"
+            BASE_PATH AS "hdfs://hdfs_host:hdfs_port/user/palo/data/input/dir/"
```
是不是可以增加partition column 声明，这样能够让用户清晰的知道自己在做什么。

imay commented:
> Add Partition Discover as a feature of Broker Load.
> 
> For usage of this feature see diffs of related docs.

@yuanlihan 
Because this patch will change load interface, I think it's better to discuss the new interface before implementation.

yuanlihan commented:
> > Add Partition Discover as a feature of Broker Load.
> > For usage of this feature see diffs of related docs.
> 
> @yuanlihan
> Because this patch will change load interface, I think it's better to discuss the new interface before implementation.

@imay 
OK. I will create an issue to discuss the interface design about this feature before updating additional diffs.

yuanlihan commented:
@imay I have added an issue about this [Enable Partition Discovery for Broker Load](https://github.com/apache/incubator-doris/issues/1582)

morningman committed:
Support multi partition column when creating table (#1574)

When creating table with OLAP engine, use can specify multi parition columns.
eg:

PARTITION BY RANGE(`date`, `id`)
(
    PARTITION `p201701_1000` VALUES LESS THAN ("2017-02-01", "1000"),
    PARTITION `p201702_2000` VALUES LESS THAN ("2017-03-01", "2000"),
    PARTITION `p201703_all`  VALUES LESS THAN ("2017-04-01")
)

Notice that load by hadoop cluster does not support multi parition column table.

morningman committed:
Fix bug that unable to delete replica if version is missing (#1585)

If there is a redundant replica on BE which version is missing,
the tablet report logic can not drop it correctly.

Commit:
Implement the initial version of BetaRowset (#1568)

yiguolei committed:
Acquire tablet map write lock during tablet gc (#1588)

imay committed:
Support Segment for BetaRowset (#1577)

We create a new segment format for BetaRowset. New format merge
data file and index file into one file. And we create a new format
for short key index. In origin code index is stored in format like
RowCusor which is not efficient to compare. Now we encode multiple
column into binary, and we assure that this binary is sorted same
with the key columns.

morningman committed:
Fix a serious bug that will cause all replicas being deleted. (#1589)

Revert commit: eda55a7394fcec2f7b6c0aefd1628f9d63911815

Youngwb committed:
Support setting timezone variable in FE (#1587)

HangyuanLiu committed:
Fix parquet directory have empty file (#1593)

worker24h committed:
Support  Decimal Type when load Parquet File (#1595)

yiguolei committed:
Manage tablet by partition id (#1591)

xy720 committed:
Merge Help document to documentation (#1586)

Help document collation (integration of help and documentation documents)

wuyunfeng committed:
Fix errors when ES username and passwd is empty (#1601)

lenmom committed:
Include header file for ‘preadv'  which caused break build on ubuntu 18.04 (#1602)

yuanlihan committed:
Enable parsing columns from file path for Broker Load

yuanlihan committed:
Update docs

yuanlihan committed:
merge conflicts
